### PR TITLE
Add azure storage endpoint suffix #26432

### LINKED ISSUE #26432: Snapshot azure repository support for Azure China
**Describe the feature**:

The azure snapshot plugin currently takes a storage account name and key as an input and tries to retrieve the CloudStorageAccount by forming a connection string via the provided input. The domain in the default case is assumed to be '**core.windows.net**' which works for most of Microsoft's data centers. 

In China, the domain is different, hence trying to store/restore a snapshot by using the storage account name and key alone as inputs does not work. The storage account access keys provide an EndpointSuffix attribute which indicates the domain to be used. As a feature, it will be nice to have an additional input attribute for the Endpoint Suffix with the default still being 'core.windows.net' if left unspecified.

We currently are running ES1.6 in china for our prod customers. In order to circumvent this, we had to include the EndpointSuffix portion as part of the account key input value since the connection string is formed through concatenation. It will be nice to have this as an explicit input.

**Elasticsearch version** (`bin/elasticsearch --version`): 1.6.2

### PR DESCRIPTION
**Pull request by liketic**
Close #26432

Happy to hear any comments to make this PR better. Thanks in advance.  😄

### CODE REVIEW

File: plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java [hunk 1]
```diff
@@ -51,6 +51,12 @@
         key -> SecureSetting.secureString(key, null));
 
     /**
+     * Azure endpoint suffix. Default to core.windows.net (CloudStorageAccount.DEFAULT_DNS).
+     */
+    public static final AffixSetting<SecureString> ENDPOINT_SUFFIX_SETTING = Setting.affixKeySetting(PREFIX, "endpoint_suffix",
```
Comment (jasontedor) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java]: Why does this need to be a secure setting?

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left a question.

File: plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java (hunk 1 above)
Comment (liketic) [reply plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java]: Thanks! I think a normal string is OK. Shall I merge with master for review?

File: plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java [hunk 2]
```diff
@@ -51,6 +51,12 @@
         key -> SecureSetting.secureString(key, null));
 
     /**
+     * Azure endpoint suffix. Default to core.windows.net (CloudStorageAccount.DEFAULT_DNS).
+     */
+    public static final Setting<String> ENDPOINT_SUFFIX_SETTING = Setting.affixKeySetting(PREFIX, "endpoint_suffix",
+        key -> SecureSetting.simpleString(key));
```
Comment (rjernst) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java]: I don't think the endpoint needs to be a secure setting.

Comment (liketic) [reply plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageSettings.java]: Yes, of course. Thanks.

**Comment by liketic** [comment]
@jasontedor @rjernst I updated the PR, please review again. Thanks!

**Comment by rjernst** [comment]
@elasticmachine please test this

**Comment by rjernst** [comment]
@elasticmachine test this

**Comment by liketic** [comment]
@rjernst Thanks, I think the testing errors are not caused by my changes...

**Comment by rjernst** [comment]
@elasticmachine retest this please

**Comment by rjernst** [comment]
I tested this locally. Looks good, thanks @liketic!

**Pull request merged**